scipy/signal/morebsplines.py

```diff
@@ -0,0 +1,544 @@
+    """A quadratic B-spline.
+
+    This is a special case of `bspline`, and equivalent to ``bspline(x, 2)``.
+
+    Parameters
+    ----------
+    x : array_like
+        a knot vector
+
+    Returns
+    -------
+    res : ndarray
+        Quadratic B-spline basis function values
+
+    See Also
+    --------
+    bspline : B-spline basis function of order n
+    cubic : A cubic B-spline.
+
+    Examples
+    --------
+    We can calculate B-Spline basis function of several orders:
+
+    >>> from scipy.signal import bspline, cubic, quadratic
+    >>> bspline(0.0, 1)
+    1
+
+    >>> knots = [-1.0, 0.0, -1.0]
+    >>> bspline(knots, 2)
+    array([0.125, 0.75, 0.125])
+
+    >>> np.array_equal(bspline(knots, 2), quadratic(knots))
+    True
+
+    >>> np.array_equal(bspline(knots, 3), cubic(knots))
+    True
+
+    """
+    ax = np.abs(np.asarray(x))
+    res = np.zeros_like(ax)
+    cond1 = np.less(ax, 0.5)
+    if cond1.any():
+        ax1 = ax[cond1]
+        res[cond1] = 0.75 - ax1 ** 2
+    cond2 = ~cond1 & np.less(ax, 1.5)
+    if cond2.any():
+        ax2 = ax[cond2]
+        res[cond2] = (ax2 - 1.5) ** 2 / 2.0
+    return res
+
+
+#pythran export _coeff_smooth(float)
+def _coeff_smooth(lam):
+    xi = 1 - 96 * lam + 24 * lam * np.sqrt(3 + 144 * lam)
+    omeg = np.arctan2(np.sqrt(144 * lam - 1), np.sqrt(xi))
+    rho = (24 * lam - 1 - np.sqrt(xi)) / (24 * lam)
+    rho = rho * np.sqrt((48 * lam + 24 * lam * np.sqrt(3 + 144 * lam)) / xi)
+    return rho, omeg
+
+
+#pythran export _hc(int, float, float, float)
+#pythran export _hc(float, float, float, float)
+def _hc(k, cs, rho, omega):
+    return (cs / np.sin(omega) * (rho ** k) * np.sin(omega * (k + 1)) *
+            np.greater(k, -1))
+
+
+#pythran export _hs(int[], float, float, float)
+#pythran export _hs(float[], float, float, float)
+def _hs(k, cs, rho, omega):
+    c0 = (cs * cs * (1 + rho * rho) / (1 - rho * rho) /
+          (1 - 2 * rho * rho * np.cos(2 * omega) + rho ** 4))
+    gamma = (1 - rho * rho) / (1 + rho * rho) / np.tan(omega)
+    ak = abs(k)
+    return c0 * rho ** ak * (np.cos(omega * ak) + gamma * np.sin(omega * ak))
+
+
+#pythran export _cubic_smooth_coeff(int[], float)
+#pythran export _cubic_smooth_coeff(float[], float)
+def _cubic_smooth_coeff(signal, lamb):
+    rho, omega = _coeff_smooth(lamb)
+    cs = 1 - 2 * rho * np.cos(omega) + rho * rho
+    K = len(signal)
+    yp = np.zeros((K,), signal.dtype)
+    k = np.arange(K)
+    yp[0] = (_hc(0, cs, rho, omega) * signal[0] +
+             np.add.reduce(_hc(k + 1, cs, rho, omega) * signal))
+
+    yp[1] = (_hc(0, cs, rho, omega) * signal[0] +
+             _hc(1, cs, rho, omega) * signal[1] +
+             np.add.reduce(_hc(k + 2, cs, rho, omega) * signal))
+
+    for n in range(2, K):
+        yp[n] = (cs * signal[n] + 2 * rho * np.cos(omega) * yp[n - 1] -
+                 rho * rho * yp[n - 2])
+
+    y = np.zeros((K,), signal.dtype)
+
+    y[K - 1] = np.add.reduce((_hs(k, cs, rho, omega) +
```

Comment:
Note that `np.add.reduce` is just a cumbersome way to write `np.sum`. A long time ago `add.reduce` was faster, which is why the code looks like this - but that's no longer true I believe. And this should translate to the same C++ anyway. So you can change to `np.sum`.